thewalrus/fock_gradients.py

```diff
@@ -23,36 +23,82 @@
 .. autosummary::
     :toctree: api
 
-    Xgate
     Dgate
     Sgate
     Rgate
     Kgate
     S2gate
     BSgate
-    Sgate_real
-    S2gate_real
-    BSgate_real
 
 """
 import numpy as np
 
 from numba import jit
 
 
-from thewalrus.libwalrus import (
-    interferometer,
-    squeezing,
-    displacement,
-    interferometer_real,
-    displacement_real,
-    squeezing_real,
-    two_mode_squeezing,
-    two_mode_squeezing_real,
-)
+@jit(nopython=True)
+def displacement_rec(alpha, cutoff):  # pragma: no cover
+    r"""Calculate the matrix elements of the real or complex displacement gate using a recursion relation.
 
+    Args:
+        alpha (float or complex): value of the displacement.
+        cutoff (int): Fock ladder cutoff
+
+    Returns:
+        (array): matrix representing the displacement operation.
+    """
+    D = np.zeros((cutoff, cutoff), dtype=np.complex128)
+    y = np.array([alpha, -np.conj(alpha)])
+    sqns = np.sqrt(np.arange(cutoff))
+
+    D[0, 0] = np.exp(-0.5 * np.abs(y[0]) ** 2)
+    D[1, 0] = y[0] * D[0, 0]
+
+    for m in range(2, cutoff):
+        D[m, 0] = y[0]/sqns[m] * D[m-1, 0]
+
+    for m in range(0, cutoff):
+        for n in range(1, cutoff):
+            D[m, n] = y[1]/sqns[n] * D[m, n-1] + sqns[m]/sqns[n] * D[m-1, n-1]
+
+    return D
+
+
+@jit(nopython=True)
+def squeezing_rec(r, theta, cutoff):  # pragma: no cover
+    r"""Calculate the matrix elements of the real or complex squeezing gate using a recursion relation.
+
+    Args:
+        r (float): amplitude of the squeezing.
+        theta (float): phase of the squeezing.
+
+    Returns:
+        (array): matrix representing the squeezing operation.
+    """
+    S = np.zeros((cutoff, cutoff), dtype=np.complex128)
+    eitheta_tanhr = np.exp(1j * theta)* np.tanh(r)
+    sinhr = 1.0 / np.cosh(r)
+    R = np.array(
+        [
+            [-eitheta_tanhr, sinhr],
+            [sinhr, np.conj(eitheta_tanhr)],
+        ]
+    )
+    sqns = np.sqrt(np.arange(cutoff))
+    S[0, 0] = np.sqrt(sinhr)
+
+    for m in range(2, cutoff, 2):
+        S[m, 0] = sqns[m-1]/ sqns[m] * R[0, 0] * S[m-2, 0]
 
-@jit("void(complex128[:,:], complex128[:,:], complex128[:,:], double)")
+    for m in range(0, cutoff):
+        for n in range(1, cutoff):
+            if (m+n)%2 == 0:
+                S[m, n] = sqns[n-1]/sqns[n] * R[1, 1] * S[m, n-2] + sqns[m]/sqns[n] * R[0, 1] * S[m-1, n-1]
+
+    return S
+
+
+@jit(nopython=True)
 def grad_Dgate(T, gradTr, gradTtheta, theta):  # pragma: no cover
```

Comment:
It can be very confusing to have functions that work as side effects. I think it's better to have the grad_xxx functions create the empty arrays and return them rather than being passed the arrays and fill them.

Reply:
This is now fixed. Thanks for pointing this out @ziofil .